### PR TITLE
DEV-5287 covid19 budget categories section

### PR DESCRIPTION
**Pull request by jameslee40**
**High level description:**
Budget Categories section added to the covid19 page.

**Technical details:**

- Created `BudgetCategories` component that holds `BudgetCategoriesCountTabContainer` and `BudgetCategoriesTableContainer`.
- Created `BaseBudgetCategoryRow` data model.
- Created `budgetCategoriesHelper` for api end points.
- Created `BudgetCategoriesTableColumns` for columns.

[Link to API CONTRACT PR](https://github.com/fedspendingtransparency/usaspending-api/pull/2519)

**JIRA Ticket:**
[DEV-5287](https://federal-spending-transparency.atlassian.net/browse/DEV-5287)

**Mockup:**
[latest mock](https://bahdigital.invisionapp.com/share/JPIAE1YWAUQ#/screens/296037726_COVID-19_Profile_-_Mid-Fidelity)

The following are ALL required for the PR to be merged:

Author:
- [x] Linked to this PR in JIRA ticket
- [x] Scheduled demo including Design/Testing/Front-end OR Provided instructions for testing in JIRA and PR `if applicable`
- [ ] Verified cross-browser compatibility: Chrome, Safari, Firefox, Internet Explorer, Edge
- [x] Verified mobile/tablet/desktop/monitor responsiveness
- [x] Verified that this PR does not create any *new* accessibility issues (via Axe Chrome extension)
- [x] Added Unit Tests for methods in Container Components, reducers, helper functions, and models `if applicable`
- [N/A ] All `componentWillReceiveProps` and `componentWillMount` in relevant child/parent components/containers replaced with [future compatible life-cycle methods](https://reactjs.org/blog/2018/03/27/update-on-async-rendering.html) per [JIRA item 3277](https://federal-spending-transparency.atlassian.net/browse/DEV-3277)
- [N/A] [API contract](https://github.com/fedspendingtransparency/usaspending-api/tree/dev/usaspending_api/api_contracts) updated `if applicable`
- [x] [Component Library Integration Status Report](https://github.com/fedspendingtransparency/data-act-documentation/blob/data-transparency-ui/frontend_apps/component-library-integration-status.md) updated `if applicable`

Reviewer(s):
- N/A Design review complete - mock has changed, will make UX edits in a follow-up PR
- N/A API merged concurrently - API integration will be in a separate PR for subtask [DEV-5394](https://federal-spending-transparency.atlassian.net/browse/DEV-5394)
- [x] Code review complete

### CODE REVIEW

File: src/_scss/pages/agency/v2/_tabs.scss [hunk 1]
```diff
@@ -7,8 +7,10 @@
         line-height: rem(27);
         text-align: center;
         > div {
-            padding: 0 rem(21) rem(30);
+            padding: rem(30) rem(15) rem(30);
+            flex: 1 0 0;
```
Comment (MaxwellKendall) [inline src/_scss/pages/agency/v2/_tabs.scss:11]: We should use the mixin from bourbon that we rely on for vendor prefixing

```suggestion
            @include flex(1 0 0);
```

File: src/_scss/pages/covid19/_budgetCategoriesTable.scss [hunk 2]
```diff
@@ -0,0 +1,70 @@
+.usda-table {
+    margin-top: 0;
+    td {
+        &:first-child {
+            @include span-columns(6 of 12, table);
```
Comment (MaxwellKendall) [inline src/_scss/pages/covid19/_budgetCategoriesTable.scss:5]: oh snap... what is this? Super nice.

File: src/_scss/pages/covid19/_budgetCategoriesTable.scss [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+            text-align: left;
+        }
+        @include span-columns(2 of 12, table);
```
Comment (MaxwellKendall) [inline src/_scss/pages/covid19/_budgetCategoriesTable.scss:8]: Nice!

File: src/_scss/pages/search/results/table/_resultsTableMessageContainer.scss [hunk 4]
```diff
@@ -0,0 +1,48 @@
+.results-table-message-container {
+    display: flex;
+    flex-flow: column wrap;
```
Comment (MaxwellKendall) [inline src/_scss/pages/search/results/table/_resultsTableMessageContainer.scss:3]: Nice! Just learned what this is.

File: src/_scss/pages/search/results/table/_resultsTableMessageContainer.scss [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    position: relative;
+    background-color: rgb(246, 246, 246);
```
Comment (MaxwellKendall) [inline src/_scss/pages/search/results/table/_resultsTableMessageContainer.scss:5]: We typically use hex values. Not sure if theres a hex defined as a variable that you might be able to use here?

File: src/js/components/covid19/budgetCategories/BudgetCategories.jsx [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+                            setActiveTab={setActiveTab}
+                            active={activeTab === tab.type}
+                            fy={2020}
+                            defCodes={defCodes} />
+                    ))}
+                </div>
+                <BudgetCategoriesTableContainer
+                    type={activeTab}
+                    subHeading={subHeading}
+                    fy={2020}
```
Comment (MaxwellKendall) [inline src/js/components/covid19/budgetCategories/BudgetCategories.jsx:81]: Obviously this is fine for now. Any thoughts on how we might make this dynamic?

File: src/js/containers/covid19/budgetCategories/BudgetCategoriesTableContainer.jsx [hunk 7]
```diff
@@ -0,0 +1,193 @@
+/**
+ * BudgetCategoriesTableContainer.jsx
+ * Created by James Lee 6/5/20
+ */
+
+import React, { useState, useEffect, useCallback } from 'react';
+import PropTypes from 'prop-types';
+import { Table, Pagination, TooltipWrapper, Picker } from 'data-transparency-ui';
```
Comment (MaxwellKendall) [inline src/js/containers/covid19/budgetCategories/BudgetCategoriesTableContainer.jsx:8]: 👀 

Nice

File: src/js/containers/covid19/budgetCategories/BudgetCategoriesTableContainer.jsx [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+    const updateSort = (field, direction) => {
+        setSort(field);
+        setOrder(direction);
+    };
+    const [results, setResults] = useState([]);
+    const [loading, setLoading] = useState(true);
+    const [error, setError] = useState(false);
+    const [spendingCategory, setSpendingCategory] = useState("total_spending");
+
+    const parseSpending = (data) => {
```
Comment (MaxwellKendall) [inline src/js/containers/covid19/budgetCategories/BudgetCategoriesTableContainer.jsx:38]:
```suggestion
    const parseSpendingDataAndSetResults = (data) => {
```

File: src/js/models/covid19/budgetCategories/BaseBudgetCategoryRow.js [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+            } else if (column.title === 'name') {
+                this[`${column.title}`] = `${data.code} — ${data.description}` || '';
+            } else {
+                defaultValue = typeof (data[column.title]) === 'string' ? '' : 0;
+                this[`${column.title}`] = data[column.title] || defaultValue;
+            }
+        });
+    },
+    populateTotalBudgetaryResources(data, type, spendingCategory) {
+        if (type !== 'object_class' && spendingCategory !== 'award_spending') {
```
Comment (MaxwellKendall) [inline src/js/models/covid19/budgetCategories/BaseBudgetCategoryRow.js:33]: This is confusing to me. Why are we only seeing the total based on type?

File: src/_scss/pages/covid19/_budgetCategoriesTable.scss (hunk 2 above)
Comment (jameslee40) [reply src/_scss/pages/covid19/_budgetCategoriesTable.scss:5]: @MaxwellKendall Yeah, it is pretty nice! It's part of bourbon apparently. https://neat.bourbon.io/docs/1.8.0/

File: src/js/components/covid19/budgetCategories/BudgetCategories.jsx (hunk 6 above)
Comment (jameslee40) [reply src/js/components/covid19/budgetCategories/BudgetCategories.jsx:81]: @MaxwellKendall If this is going to be something the user can select from a dropdown in the future, it'll be a prop passed down, but for now I'll make it into a variable.

File: src/js/models/covid19/budgetCategories/BaseBudgetCategoryRow.js (hunk 9 above)
Comment (jameslee40) [reply src/js/models/covid19/budgetCategories/BaseBudgetCategoryRow.js:33]: We're only supposed to show data for `total budgetary resources` if the `spendingCategory` is `total_spending` and when the `type` is anything other than `object_class`.

File: src/_scss/pages/search/results/table/_resultsTableMessageContainer.scss (hunk 5 above)
Comment (jameslee40) [reply src/_scss/pages/search/results/table/_resultsTableMessageContainer.scss]: Updated!

File: src/_scss/pages/covid19/_budgetCategoriesTable.scss [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+        }
+
+        &:not(:first-child) {
+            .table-header__content {
+                justify-content: flex-end;
+            }
+        }
+    }
+}
+.budget-categories-table-def_code {
```
Comment (ebdabbs) [inline src/_scss/pages/covid19/_budgetCategoriesTable.scss:30]: Looks like def code is the modifier here, so BEM class name would be `.budget-categories-table_def-code`

File: src/_scss/pages/search/results/table/_resultsTableMessageContainer.scss [hunk 11]
```diff
@@ -0,0 +1,48 @@
+.results-table-message-container {
+    display: flex;
+    flex-flow: column wrap;
+    position: relative;
+    background-color: #F6F6F6;
```
Comment (ebdabbs) [inline src/_scss/pages/search/results/table/_resultsTableMessageContainer.scss:5]: We could use the variable `$color-gray-lightest` here (to ensure color consistency throughout the site)

File: src/js/components/covid19/budgetCategories/BudgetCategories.jsx [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+    {
+        type: 'agency',
+        label: 'Agencies',
+        description: 'What agencies did the spending?',
+        countField: 'count'
+    },
+    {
+        type: 'object_class',
+        label: 'Object Classes',
+        description: 'What items or services were purchased',
```
Comment (ebdabbs) [inline src/js/components/covid19/budgetCategories/BudgetCategories.jsx:30]: Missing `?`

File: src/js/components/covid19/budgetCategories/BudgetCategoriesCountTab.jsx [hunk 13]
```diff
@@ -0,0 +1,49 @@
+/**
+ * BudgetCategoriesCountTab.jsx
```
Comment (ebdabbs) [inline src/js/components/covid19/budgetCategories/BudgetCategoriesCountTab.jsx:2]: Looks like we could reuse the `CountTab` component from agency v2. Eventually I might like to add tabs to the component library..

File: src/js/containers/covid19/budgetCategories/BudgetCategoriesTableContainer.jsx (hunk 7 above)
Comment (ebdabbs) [reply src/js/containers/covid19/budgetCategories/BudgetCategoriesTableContainer.jsx:8]: 🙌 putting the component library to good use!

File: src/js/helpers/covid19/budgetCategoriesHelper.js [hunk 14]
```diff
@@ -0,0 +1,33 @@
+/**
+ * agencyHelper.js
```
Comment (ebdabbs) [inline src/js/helpers/covid19/budgetCategoriesHelper.js:2]: budgetCategoriesHelper

File: src/js/helpers/covid19/budgetCategoriesHelper.js [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+});
+
+export const fetchDisasterSpendingCount = (type, params) => apiRequest({
+    url: `v2/disaster/${type}/count/`,
+    method: 'post',
+    data: params,
+    isMocked: true
+});
+
+export const fetchDefCodes = () => apiRequest({
```
Comment (ebdabbs) [inline src/js/helpers/covid19/budgetCategoriesHelper.js:22]: Can be removed after Jon's PR is merged - we also may consolidate all the COVID api helpers into a single file, but this is fine for now!

File: src/js/models/covid19/budgetCategories/BaseBudgetCategoryRow.js [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+ * Created by James Lee 6/05/20
+ */
+
+import { formatMoney } from 'helpers/moneyFormatter';
+import { budgetColumns, budgetColumnFields, budgetDropdownColumns, budgetFields } from 'dataMapping/covid19/budgetCategories/BudgetCategoriesTableColumns';
+
+const BaseBudgetCategoryRow = {
+    populateOnSpendingCategory(data, spendingCategory) {
+        budgetDropdownColumns[spendingCategory].forEach((column) => {
+            const defaultValue = typeof (data[budgetFields[spendingCategory][column.title]]) === 'string' ? '' : 0;
```
Comment (ebdabbs) [inline src/js/models/covid19/budgetCategories/BaseBudgetCategoryRow.js:12]: Nice!

File: src/js/models/covid19/budgetCategories/BaseBudgetCategoryRow.js [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+        return formatMoney(this._awardOutlay);
+    },
+    get totalBudgetaryResources() {
+        return formatMoney(this._totalBudgetaryResources);
+    },
+    get faceValueOfLoan() {
+        return formatMoney(this._faceValueOfLoan);
+    },
+    get countOfLoan() {
+        return this._countOfLoan;
```
Comment (ebdabbs) [inline src/js/models/covid19/budgetCategories/BaseBudgetCategoryRow.js:58]: Could we use `formatNumber` here? That will add commas for loan counts of 1000 or more

File: src/js/helpers/covid19/budgetCategoriesHelper.js (hunk 15 above)
Comment (jameslee40) [reply src/js/helpers/covid19/budgetCategoriesHelper.js:22]: Done!

File: src/js/components/covid19/budgetCategories/BudgetCategoriesCountTab.jsx (hunk 13 above)
Comment (jameslee40) [reply src/js/components/covid19/budgetCategories/BudgetCategoriesCountTab.jsx:2]: Done!

File: src/js/models/covid19/budgetCategories/BaseBudgetCategoryRow.js (hunk 17 above)
Comment (jameslee40) [reply src/js/models/covid19/budgetCategories/BaseBudgetCategoryRow.js]: Added!

File: src/js/containers/covid19/budgetCategories/BudgetCategoriesCountTabContainer.jsx [hunk 18]
```diff
@@ -21,15 +21,16 @@ const propTypes = {
...
+    const defCodes = useSelector((state) => state.covid19.defCodes);
 
     useEffect(() => {
-        if (props.defCodes) {
+        if (defCodes) {
             // Reset any existing results
             setCount(null);
 
             const params = {
                 fiscal_year: props.fy,
```
Comment (ebdabbs) [inline src/js/containers/covid19/budgetCategories/BudgetCategoriesCountTabContainer.jsx:32]: `fiscal_year` filter has been removed from the API contracts.

File: src/js/containers/covid19/budgetCategories/BudgetCategoriesTableContainer.jsx [hunk 19]
```diff
@@ -46,20 +50,20 @@ const BudgetCategoriesTableContainer = (props) => {
...
 
             if (spendingCategory === 'loan_spending') {
                 const params = {
                     filter: {
-                        def_codes: props.defCodes,
+                        def_codes: defCodes,
                         fiscal_year: props.fy
                     },
                     pagination: {
                         size: pageSize,
```
Comment (ebdabbs) [inline src/js/containers/covid19/budgetCategories/BudgetCategoriesTableContainer.jsx:63]: `size` has been changed to `limit`

**Comment by ebdabbs** [comment]
There have been a couple of changes to the mockup and some upcoming API contract changes, but going to go ahead and merge what we have!

**Pull request merged**